Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +19,53 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return ampHourUsed;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+   return isCharging_;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    if(isCharging_){
+        return holdC;
+    }else{
+        return holdD;
+
+    }
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
+   // Q_UNUSED(batteryData);
     // This is where you can update your variables
     // Hint: There are many different ways that the totalAmpHoursUsed can be updated
     // i.e: Taking a running average of your data values, using most recent data points, etc.
+    if(batteryData.current<0){
+        isCharging_=true;
+    }else{
+        isCharging_=false;
+    }
+
+    if(ampHourUsed==0.0){
+        ampHourUsed=BATTERY_AMP_HOUR_CAPACITY*(initialStateOfChargePercent_/100);
+    }
+    else{
+        ampHourUsed+=((prevCurr+batteryData.current)/2)*((prevTime.msecsTo(batteryData.time))/3600000.0) ;
```

Comment:
Change 3600000.0 to constant

Reply:
For further elaboration, the value `3600000.0` is something more of a "Magic Number". Just looking at the value, it's not immediately obvious that it represents the number of seconds in an hour. 

Typical convention is to put it in a constant value in a namespace, all caps: 
`const double MY_CONSTANT_VALUE = 3600000.0"